java/maven/src/org/netbeans/modules/maven/NbMavenProjectImpl.java

```diff
@@ -559,29 +587,123 @@ void stopHardReferencingMavenPoject() {
         return newproject;
     }
 
+    /**
+     * Task that potential project reloads should wait on. If set, a {@link fireProjectReload}(true) will be scheduled only after this blocker finishes.
+     */
+    private RequestProcessor.Task reloadBlocker;
 
-
+    /**
+     * Schedules project operation that delays potential reloads. If a reload is posted, it will be performed only after
+     * this operation compeltes (successfully, or erroneously). Multiple project operations can be scheduled, an eventual project reload
+     * should happen after all those operations complete. It is possible to postpone project reload indefinitely, avoid unnecessary
+     * operation schedules.
+     * 
+     * @param r operation to run
+     * @return controlling task.
+     */
+    public RequestProcessor.Task scheduleProjectOperation(RequestProcessor.Task t) {
+        if (Boolean.getBoolean("test.reload.sync")) {
+            LOG.log(Level.FINE, "Running the blocking task synchronously (test.reload.sync set)");
+            t.run();
+            return t;
+        } else {
+            synchronized (this) {
+                if (reloadBlocker == null) {
+                    LOG.log(Level.FINER, "Blocking project reload on task {0}", t);
+                    reloadBlocker = t;
+                    return t;
+                } else {
```

Comment:
this looks a bit dangerous, the task is already running when passed into this method, which means that there is a point in time when the task is active without being marked as blocker.
```
((NbMavenProjectImpl)this.nbproject).scheduleProjectOperation(MavenModelProblemsProvider.RP.post(toRet));
```
Should this method have a RP param and actually schedule the tasks _inside_ the method? (+ throw exception when the task is already running).

Reply:
Good point. I'm glad that this is not a public API ... will follow the advice.